### DISCUSSION TITLE
Same index with different values

### DISCUSSION THREAD

**Discussion by Stocki916** [ravendb/ravendb#14843] [General]
Dear community,

we are using a database with an index which calculates same data (average price per month). After we cloned this index without any other modifications, the computed values differ in both indizes and show different values. How is this possible with the same logic and the same data which is used?
Both indizes have the same count of items but different values:

Index 1:
![image](https://user-images.githubusercontent.com/60972798/187912529-acfaccd1-912f-4aef-8d8d-0ae2d0fc6295.png)

Index 2:
![image](https://user-images.githubusercontent.com/60972798/187912589-869ff5a0-2a20-4075-9ab3-e26cf0ba8661.png)

Any ideas how we could check what happens here?

**Comment by ayende** [comment]
Is it possible that you have an issue like the one shown here:

https://ayende.com/blog/197637-A/production-postmortem-your-math-is-wrong-recursion-doesnt-work-this-way

**Reply by Stocki916** [reply]
Thanks for your fast reply! I checked your blog post but we don't aggregate data using multiple maps, only one single map and calculate the average price.
Here are some screenshots of the index definition and also the results of the map reduce visualizer:
![MicrosoftTeams-image](https://user-images.githubusercontent.com/60972798/188071499-a625ede5-2cb5-48a6-9a02-33c934d78a21.png)


In the map reduce visualizer there are different counts of data in the leafs, but the total count at top level is identical (except the calculated average price). But I'm not sure how to interprete this result:

![MicrosoftTeams-image (1)](https://user-images.githubusercontent.com/60972798/188071188-c3e6964a-4103-4ef7-847f-da9ddf239d74.png)
![MicrosoftTeams-image (2)](https://user-images.githubusercontent.com/60972798/188071197-62f2b1ac-586f-4403-afdd-93b71f326b69.png)

**Reply by ayende** [reply]
Yes, that is the issue. You cannot do that :-)

Average is not reenterant. When you aggregate it again on the next step, you are doing an average of averages.

What you need to do is to have an index like this:

```
from o in docs.Orders
select new 
{
   OrdersCount = 1,
   TotalPrice = o.Price,
   AveragePrice = o.Price,
   o.Date.Year
}

from result in results
group result by result.Year into g
let sum= new {
   TotalPrice=   g.Sum(x=>x.TotalPrice)
   OrdersCount= g.Sum(x=>x.OrdersCount)
}
select new 
{
   Year = g.Key,
   sum.TotalPrice,
   sum.OrdersCount,
   AveragePrice = sum.TotalPrice / sum.OrdersCount
   
}
```

In this way, you aren't trying to do average of average, but compute it from the fields where you can use recursive values.

**Reply by Stocki916** [reply]
Hmm, i checked the index again, but i think its strange, that  "PriceCount" on the leaf level contains a sum of 24k EUR (all prices in the leaf are summed up) but on aggregation level its only 19k EUR. Why is it less on top level, than on the leaf below? On top level i would excepct > 4 mio EUR but not less than on the leafs below.

![MicrosoftTeams-image (3)](https://user-images.githubusercontent.com/60972798/188450408-72e61646-4738-4525-aaa7-add94d7e6c71.png)

The total count of price values ("Count") seems to be correct.
I refactored the index before to be like this, but is still strange (as mentioned above):
![MicrosoftTeams-image (4)](https://user-images.githubusercontent.com/60972798/188451121-d4c7acf5-311f-4d8a-a208-333882be44f2.png)

**Reply by ayende** [reply]
You are looking at a _single_ leaf, but note:
![image](https://user-images.githubusercontent.com/116915/188451817-d5209c98-73e1-4349-b2d1-32965373c8cf.png)


There are many others that aren't visible for you.

**Reply by ayende** [reply]
They also contain data, and the total average is 19K, I assume.

**Reply by Stocki916** [reply]
yes that what i mean. If this single node in the middle already contains prices with a sum of "24.605,88" - why is the total sum above (including also the other leafes left and right) lower?
That can't be true because we have no negativ prices :-)

**Reply by ayende** [reply]
See: `TotalPrice = g.Sum(x=>x.PriceValue)`

However, `PriceValue` is the _average_, I think you need to change that to:

`TotalPrice = g.Sum(x=>x.PriceCount)` <-- note that this is a very mislead name. Would suggest renaming that to `TotalPrice`

**Comment by Stocki916** [comment]
For it seems more like a not thread safe loop internally? "PriceValue" is no average, its just the total sum of all prices. The "PriceValue" of each "PriceItem" is summed up and saved to "TotalPrice". Each "PriceItem" has also a value called "PriceCount" with value 1 which is also summed up and stored in "TotalCount" on top level.
We also removed the "PriceCount" property and just used the query "Count()" to count all items in the grouping, but with no effect.

The average and counted values on the leaf seems to be correct, but total values on top are always wrong. May it be, that there is a not "thread safe" loop in the background, which collects all "sub-results" from the leafs?

If you like we could also start a remote session and demonstrate the issue directly

Here we tested with a simplified index, but also with wrong result:

![MicrosoftTeams-image (5)](https://user-images.githubusercontent.com/60972798/188648121-9181d7bd-e202-4e01-9d84-66d4540d64b7.png)

![MicrosoftTeams-image (6)](https://user-images.githubusercontent.com/60972798/188648161-4dca5880-7872-40b0-882e-378ebb20dafb.png)

**Reply by ayende** [reply]
When you post code, please post it as text, that makes it much easier to work with.

There is not thread safety here.
There is reentracy issue, however.
Look at the the `PriceValue` you have here.
That isn't the *actual* sum, that is the average you have there.
And when you recurse into that, it won't give the right results.

Consider the following data:

```
// entries from the map
{PriceCount = 1, PriceValue = 7}
{PriceCount = 1, PriceValue = 5}
```

Now, let's run this through the reduce, you'll get:

TotalPrice = Sum(5,6)

And the output would be:

`{PriceCount = 2, PriceValue = 6}`

Now... let's add a document, where the map would be:

`{PriceCount = 1, PriceValue = 3}`

This is then fed into the reduce with the following inputs:

`Reduce( [
   {PriceCount = 2, PriceValue = 6}, // the previous reduced results
   {PriceCount = 1, PriceValue = 3}  // the new mapped value
])`


What would your code generate?

Well, you are using `g.Count()`, which is 2, and you are computing the sum of `PriceValue`.
However, `PriceValue` from the previously reduced results is `6`, so your output given your code would be:

`{PriceCount = 2, PriceValue = 4.6}`

What is the problem?
You _cannot_ use `g.Count()`, and you *must* carry the total sums.

You need another set of fields, like the ones I originally shown.
Your reduce should look like:

```
from r in results
group r by new { ... } into g
let sum = new 
{
   TotalPrice = g.Sum( x => x.TotalPrice), // IMPORTANT, **NOT** PriceValue
   TotalCount = g.Sum( x => x.TotalCount), // IMPORTNANT, **NOT** Count()
}
select new
{
   ...
   PriceValue = sum.TotalPrice / sum.TotalCount, // we never aggergate on this further!
   sum.TotalPrice,
   sum.TotalCount
}
```

**Reply by Stocki916** [reply]
I don't get your example:

> // entries from the map
{PriceCount = 1, PriceValue = 7}
{PriceCount = 1, PriceValue = 5}

Ok, clear so far :-)

> TotalPrice = Sum(5,6) --> Why? "TotalPrice" should be the sum if "PriceValues", so "7" + "5" = "12" and not "6"

> And the output would be:

> {PriceCount = 2, PriceValue = 6} --> No, it should be "PriceCount=2" and "PriceValue=12"

I would expect "sum" to get the sum up of the values.

Also, if the values on the leafs are correct, why are they so different on the top level?
Maybe it helps if I create a support ticket and we can start a remote session?

**Reply by ayende** [reply]
Look at the `Reduce` portion. You feed it those two values, and then you have.

`sum = new { TotalPrice = s.Sum(PriceValue) }`  - which is 12.
But that isn't the end.
You also have:

`PriceValue = sum.TotalPrice / g.Count()` - value of that is 6. 

You aren't account for the division there, no?

**Reply by Stocki916** [reply]
I don't get your explanation and think we are talking past each other ;-)
Is it possible to have a call / remote session to demonstrate it directly, or shall i create a support ticket for better handling?

**Reply by garayx** [reply]
> I don't get your example:
> 
> > // entries from the map
> > {PriceCount = 1, PriceValue = 7}
> > {PriceCount = 1, PriceValue = 5}
> 
> Ok, clear so far :-)
> 
> > TotalPrice = Sum(5,6) --> Why? "TotalPrice" should be the sum if "PriceValues", so "7" + "5" = "12" and not "6"

There is a typo TotalPrice = Sum(5,**7**), its a sum method so its value is indeed "7" + "5" = "12"
> 
> > And the output would be:
> 
> > {PriceCount = 2, PriceValue = 6} --> No, it should be "PriceCount=2" and PriceValue=12""PriceValue=12"

why `PriceValue` is 12 ?  you divide `sum.TotalPrice` (12) by `g.Count()` (which is 2), see:

![image](https://user-images.githubusercontent.com/16625387/189865577-5e554d01-a612-41c7-8ea7-fc1dda635c5e.png)

> I would expect "sum" to get the sum up of the values.
The sums is correct

> Also, if the values on the leafs are correct, why are they so different on the top level?

The index is not correct, you cannot use `g.Average()` and `g.Count()` (because they are intermediate reduce results) to calculate the average, you have to do as it is was explained in reply above: https://github.com/ravendb/ravendb/discussions/14843#discussioncomment-3533305

I am attaching you another reply with more explanations of map-reduce indexing process in RavenDB, I hope it will make the details clear.

**Comment by garayx** [comment]
The map-reduce indexing is a process where the reduce phase isn't a single operation. Note that a change to a single document might affect the aggregation results of map-reduce index. In order to avoid recalculating entire data set when a single document is changed, the outputs of reduce function are put into "buckets" (actually they are put into b+tree but it is an implementation detail). Then when a document is changed we don't need to recalculate entire data set but only a few relevant buckets. It means the reduce function is supposed to be called multiple times when calculating a final aggregation result for a given reduce key.

It means that a reduce function needs to be given in a way that it returns consistent results when it's called multiple times - for each bucket.In you case it means you need to maintain the original values from the map phase, not to rely on the values of intermediate reduce results.

The most obvious example is how to calculate the average. The issue is described here:

https://ravendb.net/learn/inside-ravendb-book/reader/4.0/11-mapreduce-and-aggregations-in-ravendb#common-pitfalls-with-mapreduce-indexes

(Actually I recommend reading entire chapter about map-reduce indexes. It's great reading - https://ravendb.net/learn/inside-ravendb-book/reader/4.0/11-mapreduce-and-aggregations-in-ravendb)

Regarding the design of map-reduce mechanism you can read here:

https://ayende.com/blog/174051/the-design-of-ravendb-4-0-you-cant-see-the-map-reduce-from-all-the-trees
https://ayende.com/blog/179938/ravendb-4-0-unsung-heroes-map-reduce

Also in order to see intermediate results of reduce phase you can use Map-Reduce Visualizer:
https://ravendb.net/docs/article-page/4.2/csharp/studio/database/indexes/map-reduce-visualizer

To conclude, the main point is that the reduce phase is multiple stages process. So the reduce function needs to take it into account.
